firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalDocumentsView.java

```diff
@@ -251,16 +260,91 @@ void recalculateAndSaveOverlays(Set<DocumentKey> documentKeys) {
     return results;
   }
 
-  /** Queries the remote documents and overlays by doing a full collection scan. */
+  /**
+   * Given a collection group, returns the next documents that follow the provided offset, along
+   * with an updated offset.
+   *
+   * <p>This method returns documents in order of read time from the provided offset, followed by
+   * documents with mutations in order of batch id from the offset. Since all documents in a batch
+   * are returned together, the total number of documents returned can exceed {@code count}.
+   *
+   * <p>If no documents are found, returns an empty list and an offset with the latest read time in
+   * the remote document cache.
+   *
+   * @param collectionGroup The collection group for the documents.
+   * @param offset The offset to index into.
+   * @param count The number of documents to return
+   * @return A pair containing the next offset and a list of documents that follow the provided
+   *     offset.
+   */
+  public Pair<IndexOffset, ImmutableSortedMap<DocumentKey, Document>> getNextDocumentsAndOffset(
+      String collectionGroup, IndexOffset offset, int count) {
+    ImmutableSortedMap<DocumentKey, Document> documents =
+        getDocuments(collectionGroup, offset, count);
+
+    // TODO(indexing): Store largest batch id on Document class so we can avoid fetching documents
+    // a second time.
+    ImmutableSortedMap<Integer, List<Document>> batchIdToDocuments =
+        getDocumentsBatchIdsMatchingCollectionGroupQuery(collectionGroup, offset);
+    IndexOffset newOffset = getNewOffset(documents, offset);
+
+    // Start updating by mutation batch id if under limit.
+    while (documents.size() < count && !batchIdToDocuments.isEmpty()) {
+      int lowestBatchId = batchIdToDocuments.getMinKey();
+      List<Document> documentsInBatch = batchIdToDocuments.get(lowestBatchId);
+
+      for (Document document : documentsInBatch) {
+        documents = documents.insert(document.getKey(), document);
+      }
+      newOffset =
+          IndexOffset.create(newOffset.getReadTime(), newOffset.getDocumentKey(), lowestBatchId);
+      batchIdToDocuments = batchIdToDocuments.remove(lowestBatchId);
+    }
+
+    return new Pair<>(newOffset, documents);
+  }
+
+  /**
+   * Returns a mapping of batch id to documents that match the provided query sorted by batch id
+   * ascending order.
+   */
+  // TODO(indexing): Support adding local mutations to collection group table.
+  public ImmutableSortedMap<Integer, List<Document>>
+      getDocumentsBatchIdsMatchingCollectionGroupQuery(String collectionGroup, IndexOffset offset) {
```

Comment:
This name is quite the handful but it should go away once we add the batch ID to the `Document` class.